### PR TITLE
[FLINK-4702] [kafka connector] Commit offets to Kafka asynchronously

### PR DESCRIPTION
**Pull request by StephanEwen**
The offset commit calls to Kafka may occasionally take very long. In that case, the notifyCheckpointComplete() method blocks for long and the KafkaConsumer cannot make progress and cannot perform checkpoints.

This pull request changes the offset committing to use Kafka's `commitAsync()` method.
It also makes sure that no more than one commit is concurrently in progress, to that commit requests do not pile up.

### CODE REVIEW

**Comment by StephanEwen** [comment]
@robert and @tzulitai What is your take on this?

**Comment by tzulitai** [comment]
Just had a look at the API of `commitAsync`, and it seems like the committed offsets back to Kafka through this API (likewise for `commitSync`) need to be `lastProcessedMessageOffset + 1` ([https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#commitAsync(java.util.Map,%20org.apache.kafka.clients.consumer.OffsetCommitCallback)](https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#commitAsync(java.util.Map,%20org.apache.kafka.clients.consumer.OffsetCommitCallback))).

This mainly effects that when starting from group offsets in Kafka, `FlinkKafkaConsumer09` currently starts from the wrong offset. There's a separate JIRA for this bug: [FLINK-4618](https://issues.apache.org/jira/browse/FLINK-4618).

Another contributor had already picked up FLINK-4618, so I'd say it's ok to leave this PR as it is. I'll help check on FLINK-4618 progress and make sure it gets merged after this PR.

Minus the above, this looks good to me. +1

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java [hunk 1]
```diff
@@ -86,6 +90,9 @@
 	/** Flag to mark the main work loop as alive */
 	private volatile boolean running = true;
 
+	/** Flag indicating whether a commit of offsets to Kafka it currently happening */
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: nit: it --> "is"?

**Comment by tzulitai** [comment]
Btw, just curious, does 0.8 Kafka connector have the same issue with sync committing? I haven't looked into the code for this, but just wondering if we need a ticket for 0.8 too.

**Comment by tzulitai** [comment]
@StephanEwen I think you've tagged the wrong Github ID for Robert ;)

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java [hunk 2]
```diff
@@ -301,4 +316,16 @@ public void commitSpecificOffsetsToKafka(Map<KafkaTopicPartition, Long> offsets)
...
 	}
+
+	private class CommitCallback implements OffsetCommitCallback {
+
+		@Override
+		public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+			commitInProgress = false;
+
+			if (exception != null) {
+				LOG.warn("Committing offsets to Kafka failed. This does not compromise Flink's checkpoints", exception);
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: The exception message isn't included in the log warning.

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java [hunk 3]
```diff
@@ -285,7 +293,14 @@ public void commitSpecificOffsetsToKafka(Map<KafkaTopicPartition, Long> offsets)
 
 		if (this.consumer != null) {
 			synchronized (consumerLock) {
-				this.consumer.commitSync(offsetsToCommit);
+				if (!commitInProgress) {
+					commitInProgress = true;
+					this.consumer.commitAsync(offsetsToCommit, offsetCommitCallback);
+				}
+				else {
+					LOG.warn("Committing previous checkpoint's offsets to Kafka not completed. " +
```
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: If the user sets a relatively short checkpoint interval, will this be flooding log?

**Comment by StephanEwen** [comment]
Thanks @tzulitai for looking at this. I will leave the offset then as it is (fixed via followup) and 

The Kafka 0.8 connector needs a similar change. This here is encountered by a user, so I wanted to get the 0.9 fix in faster. Will do a follow-up for Kafka 0.8. Will also correct the issue tag ;-)

I have no good idea how to test this, though, so any thoughts there are welcome!

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java (hunk 3 above)
Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: Possibly yes. But on the other hand, this should be pretty visible if it happens.
I would expect that with proper options to participate in group checkpoint committing, most Flink jobs run without committing to Kafka/ZooKeeper.

**Comment by tzulitai** [comment]
@StephanEwen 
On a second look, I think the `commitSpecificOffsetsToKafka` method was designed to commit synchronously in the first place. `AbstractFetcher` holds a Map of all current pending offsets for committing by checkpointID, and on every `notifyCheckpointComplete` the offsets are removed from the Map before `commitSpecificOffsetsToKafka` is called.

So, for async committing, I think we need to remove cleaning up the offsets in `AbstractFetcher#notifyCheckpointComplete()` and instead clean them up in a new separate callback handle method in `AbstractFetcher`.

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java (hunk 3 above)
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: I agree, makes sense.

**Comment by tzulitai** [comment]
Seems like currently only the 0.8 Kafka connector have tests related to offset committing (in `Kafka08ITCase`).

My two cents for testing this for now is that a IT test for correct offset committing back to Kafka in the 0.9 connector is sufficient (can take a look at `Kafka08ITCase#testOffsetInZookeeper`, but replacing `ZookeeperOffsetHandler` with the new `KafkaConsumer` methods).

File: flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java (hunk 2 above)
Comment (tzulitai) [inline flink-streaming-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java]: Oops, this is actually correct, sorry.

**Comment by StephanEwen** [comment]
@tzulitai Thanks for thorough review!

I don't understand the problem why the `commitSpecificOffsetsToKafka` method is designed to commit synchronously. The `FlinkKafkaConsumerBase` has the pending checkpoints (I think that is what you refer to). It removes the HashMap of "offsets to commit" from the `pendingCheckpoints` Map synchronously, before even calling the fetcher to commit.
After that, it looks to me like it does not make a difference how that Map "offsets to commit" is used (sync or async)...

**Comment by StephanEwen** [comment]
Actually, just discovered that the problem is different all together.

While the KafkaConsumer is polling for new data (with a timeout), it holds the consumer lock. If no data comes in Kafka, the lock is not released before the poll timeout is over.
During that time, neither a "commitSync" nor "commitAsync" call can be fired off. The `notifyCheckpointComplete` method hence blocks until the poll timeout is over and the lock is released.

We can fix this by making sure that the consumer is "woken up" to release the lock, and by making sure the lock acquisition is fair, so the committer will get it next.

For the sake of releasing the lock fast in the committer method, it should still be an asynchronous commit.

**Comment by StephanEwen** [comment]
Closing this for #2574

**Pull request closed without merging**